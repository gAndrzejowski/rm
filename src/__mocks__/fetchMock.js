export default async uri => Promise.resolve({
  json: async () => Promise.resolve({ data: uri }),
});
