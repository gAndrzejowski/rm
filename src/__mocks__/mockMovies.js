const mockMovies = [
    {
        "id": 185567,
        "title": "Zulu",
        "tagline": "",
        "vote_average": 6.7,
        "vote_count": 200,
        "release_date": "2013-12-04",
        "poster_path": "https://image.tmdb.org/t/p/w500/xg7Dh7mjevDgznqw9JHYccNo9ZQ.jpg",
        "overview": "As a child, Ali Neuman narrowly escaped being murdered by Inkhata, a militant political party at war with Nelson Mandela's African National Congress. Only he and his mother survived the carnage of those years. But as with many survivors, the psychological scars remain.",
        "budget": 16000000,
        "revenue": 0,
        "genres": [
            "Crime",
            "Drama",
            "Thriller"
        ],
        "runtime": 110
    },
    {
        "id": 269149,
        "title": "Zootopia",
        "tagline": "Welcome to the urban jungle.",
        "vote_average": 7.7,
        "vote_count": 6795,
        "release_date": "2016-02-11",
        "poster_path": "https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg",
        "overview": "Determined to prove herself, Officer Judy Hopps, the first bunny on Zootopia's police force, jumps at the chance to crack her first case - even if it means partnering with scam-artist fox Nick Wilde to solve the mystery.",
        "budget": 150000000,
        "revenue": 1023784195,
        "genres": [
            "Animation",
            "Adventure",
            "Family",
            "Comedy"
        ],
        "runtime": 108
    },
    {
        "id": 19908,
        "title": "Zombieland",
        "tagline": "This place is so dead",
        "vote_average": 7.2,
        "vote_count": 4509,
        "release_date": "2009-10-07",
        "poster_path": "https://image.tmdb.org/t/p/w500/vUzzDpVrab1BOG3ogxhRGfLN94d.jpg",
        "overview": "Columbus has made a habit of running from what scares him. Tallahassee doesn't have fears. If he did, he'd kick their ever-living ass. In a world overrun by zombies, these two are perfectly evolved survivors. But now, they're about to stare down the most terrifying prospect of all: each other.",
        "budget": 23600000,
        "revenue": 102391382,
        "genres": [
            "Comedy",
            "Horror"
        ],
        "runtime": 88
    },
    {
        "id": 1949,
        "title": "Zodiac",
        "tagline": "There's more than one way to lose your life to a killer.",
        "vote_average": 7.3,
        "vote_count": 2780,
        "release_date": "2007-03-02",
        "poster_path": "https://image.tmdb.org/t/p/w500/bgLyOROfFQI3FqYL7jQbiaV8lkN.jpg",
        "overview": "The true story of the investigation of the \"Zodiac Killer\", a serial killer who terrified the San Francisco Bay Area, taunting police with his ciphers and letters.  The case becomes an obsession for four men as their lives and careers are built and destroyed by the endless trail of clues.",
        "budget": 65000000,
        "revenue": 84785914,
        "genres": [
            "Crime",
            "Drama",
            "Mystery",
            "Thriller"
        ],
        "runtime": 157
    },
    {
        "id": 97630,
        "title": "Zero Dark Thirty",
        "tagline": "The Greatest Manhunt in History",
        "vote_average": 6.7,
        "vote_count": 2028,
        "release_date": "2012-12-19",
        "poster_path": "https://image.tmdb.org/t/p/w500/yg6IDNucLAEj7E5loTyTnUW2pgb.jpg",
        "overview": "A chronicle of the decade-long hunt for al-Qaeda terrorist leader Osama bin Laden after the September 2001 attacks, and his death at the hands of the Navy S.E.A.L. Team 6 in May, 2011.",
        "budget": 40000000,
        "revenue": 132820716,
        "genres": [
            "Thriller",
            "Drama",
            "History"
        ],
        "runtime": 157
    },
    {
        "id": 6795,
        "title": "Zathura: A Space Adventure",
        "tagline": "Adventure Is Waiting",
        "vote_average": 6.1,
        "vote_count": 1080,
        "release_date": "2005-11-06",
        "poster_path": "https://image.tmdb.org/t/p/w500/amqgIuISRBt8tsZM6cTT6gO9WLR.jpg",
        "overview": "After their father is called into work, two young boys, Walter and Danny, are left in the care of their teenage sister, Lisa, and told they must stay inside. Walter and Danny, who anticipate a boring day, are shocked when they begin playing Zathura, a space-themed board game, which they realize has mystical powers when their house is shot into space. With the help of an astronaut, the boys attempt to return home.",
        "budget": 65000000,
        "revenue": 64321501,
        "genres": [
            "Family",
            "Fantasy",
            "Science Fiction",
            "Adventure"
        ],
        "runtime": 101
    },
    {
        "id": 133931,
        "title": "Zambezia",
        "tagline": "Airborne to be wild.",
        "vote_average": 5.2,
        "vote_count": 89,
        "release_date": "2012-06-05",
        "poster_path": "https://image.tmdb.org/t/p/w500/dLZTSTe4czdJ3Fr6vwy8XwVzLF5.jpg",
        "overview": "Set in a bustling bird city on the edge of the majestic Victoria Falls, \"Zambezia\" is the story of Kai - a na√Øve, but high-spirited young falcon who travels to the bird city of \"Zambezia\" where he discovers the truth about his origins and, in defending the city, learns how to be part of a community",
        "budget": 20000000,
        "revenue": 0,
        "genres": [
            "Comedy",
            "Animation",
            "Adventure",
            "Family"
        ],
        "runtime": 83
    },
    {
        "id": 10358,
        "title": "Zack and Miri Make a Porno",
        "tagline": "What would you do to get out of debt?",
        "vote_average": 6.1,
        "vote_count": 894,
        "release_date": "2008-09-18",
        "poster_path": "https://image.tmdb.org/t/p/w500/zxwEJBhSSu5AuKTTd4ugWaCqmg.jpg",
        "overview": "Lifelong platonic friends Zack and Miri look to solve their respective cash-flow problems by making an adult film together. As the cameras roll, however, the duo begin to sense that they may have more feelings for each other than they previously thought.",
        "budget": 24000000,
        "revenue": 42105111,
        "genres": [
            "Comedy"
        ],
        "runtime": 102
    }
]
export default mockMovies;