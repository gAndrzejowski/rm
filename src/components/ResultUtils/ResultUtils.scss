.resultUtils {
  background:wheat;
  display:flex;
  justify-content: space-between;
  padding:0.5em 5vw;
  p, div {
    display:flex;
    justify-content: space-around;
    align-items:center;
    color:black;
    font-weight: bold;
  }
  .sort {
    width:20vw;
  }
}
.active {
  color:firebrick!important;
}
