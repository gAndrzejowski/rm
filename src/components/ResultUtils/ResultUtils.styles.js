const styles = {
  resultUtils: {
    background: 'wheat',
    display: 'flex',
    justifyContent: 'space-between',
    padding: ['0.5em', '5vw'],
    '@global p, div': {
      display: 'flex',
      justifyContent: 'space-around',
      alignItems: 'center',
      color: 'black',
      fontWeight: 'bold',
    },
  },
  sort: {
    width: '20vw',
    '@global div': {
      cursor: 'pointer',
    },
  },
  active: {
    color: ['firebrick', '!important'],
  },
};
export default styles;

