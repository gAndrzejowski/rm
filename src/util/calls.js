export const swagger = async settings => {
    const {
        sortBy,
        sortOrder,
        search,
        searchBy,
        filter,
    } = settings;
    const response = await fetch(`http://react-cdp-api.herokuapp.com/movies?search=${search || ''}&searchBy=${searchBy || 'title'}${filter ? ('&filter=' + filter) : ''}&sortBy=${sortBy || 'title'}&sortOrder=${sortOrder || 'desc'}`);
    return await response.json();
};

export default swagger;