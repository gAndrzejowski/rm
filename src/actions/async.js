import swagger, {single} from '../util/calls';
import {get_movies, set_current_movie, sort_movies} from "./creators";

export const searchMovies = async (query, searchBy) => {
    const results = await swagger({
        search: query,
        searchBy
    });
    return get_movies(results.data);
};

export const searchAndSort = async (sortBy, search, searchBy) => {
    const results = await swagger({
        sortBy,
        search,
        searchBy
    });
    console.log(results.data.length);
    return sort_movies(sortBy, results.data);
};

export const getByGenre = async (genre) => {
    const results = await swagger({
        search: genre,
        searchBy: "genres"
    });
    return get_movies(results.data);
};

export const getById = async id => {
    const results = await single(id);
    return set_current_movie(results);
};

export default searchMovies;