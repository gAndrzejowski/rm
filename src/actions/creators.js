import actions from './names';

const {
    GET_MOVIES,
    SORT_MOVIES,
    SET_SEARCH_CRITERION,
    SET_CURRENT_MOVIE,
    SET_SEARCH_TXT,
    HYDRATE_STORE
} = actions;

export const get_movies = (payload) => ({type: GET_MOVIES, payload});
export const sort_movies = (by, results) => ({type: SORT_MOVIES, by, results});
export const set_search_criterion = (criterion) => ({type: SET_SEARCH_CRITERION, criterion});
export const set_current_movie = (movie) => ({type: SET_CURRENT_MOVIE, movie});
export const set_search_txt = (txt)=> ({type: SET_SEARCH_TXT, txt});
export const hydrate_store = () => {
    const retrieval = (window && window.localStorage)?
        JSON.parse(window.localStorage.getItem('netflixRouletteStore')):
        null;
    return {
        type: HYDRATE_STORE,
        retrievedStore: retrieval || {}
    };
};

export default {
    get_movies,
    sort_movies,
    set_current_movie,
    set_search_criterion,
    set_search_txt,
    hydrate_store
};