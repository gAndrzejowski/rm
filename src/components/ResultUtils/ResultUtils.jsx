// @flow
import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import injectStylesheet from 'react-jss';
import { searchAndSort } from '../../actions/async';
import styles from './ResultUtils.styles';
import type { MovieData, Store } from '../../flowTypes';

type Props = {
  numFound: number,
  currentSort: string,
  search: string,
  by: string,
  changeSorting: (string, string, string) => Promise<any>,
  selected: MovieData,
  classes: Object,
}

export const ResultUtils = ({
  numFound, currentSort, changeSorting, search, by, selected, classes,
} :Props) => (
  <div className={classes.resultUtils}>
    {numFound > 0 && (
    <Fragment>
      <p>{(selected && selected.genres) ? `Films by ${selected.genres[0]} genre` : `${numFound} movies found`}</p>
      <div className={classes.sort}>
        <div>Sort By</div>
        <div onClick={() => changeSorting('release_date', search, by)} className={currentSort === 'release_date' ? classes.active : ''} role="button">release date</div>
        <div onClick={() => changeSorting('vote_average', search, by)} className={currentSort === 'vote_average' ? classes.active : ''} role="button">rating</div>
      </div>
    </Fragment>
            )}
  </div>
);
ResultUtils.defaultProps = {
  classes: {},
};

export const mapStateToProps = (state :Store) => ({
  numFound: state.movies.results.length,
  currentSort: state.movies.sort,
  search: state.heading.search,
  by: state.heading.by,
  selected: state.heading.selected,
});
export const mapDispatchToProps = (dispatch :any) => ({
  changeSorting: async (crit :string, search :string, by :string) =>
    dispatch(await searchAndSort(crit, search, by)),
});

export default connect(mapStateToProps, mapDispatchToProps)(injectStylesheet(styles)(ResultUtils));
