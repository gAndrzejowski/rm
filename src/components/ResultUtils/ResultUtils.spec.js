import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { ResultUtils, mapDispatchToProps, mapStateToProps } from './ResultUtils';
import styles from 'ResultUtils.scss';
import fetchMock from '../../__mocks__/fetchMock';
import { sort_movies } from "../../actions/creators";
import { swaggerBase, swagger } from "../../util/calls";
import mockMovies from '../../__mocks__/mockMovies';

describe ('ResultUtil component', () => {
    it('renders correctly with movie not selected', () => {
        const tree = renderer.create(<ResultUtils changeSorting={jest.fn()} currentSort="release_date" numFound={10} />);
        expect(tree).toMatchSnapshot();
    });
    it('renders correctly with movie selected', () => {
        const tree = renderer.create(<ResultUtils changeSorting={jest.fn()} selected={mockMovies[0]} currentSort="release_date" numFound={5} />)
        expect(tree).toMatchSnapshot();
    });
    it('assigns active class to proper option div based on currentSort prop', () => {
        const wrapper = shallow(<ResultUtils changeSorting={jest.fn()} currentSort="release_date" numFound={10} />)
        expect(wrapper.find(`.${styles.active}`).text()).toBe('release date');

        wrapper.setProps({currentSort: 'vote_average'});
        expect(wrapper.find(`.${styles.active}`).text()).toBe('rating');
    });
    it('calls changeSorting with proper args when one of the option divs is clicked', () => {
        const wrapper = shallow(<div><ResultUtils changeSorting={jest.fn()} currentSort="release_date" numFound={10} /></div>);
        const utils = wrapper.find('ResultUtils').first().dive();
        utils.find('div[onClick]').first().simulate('click');
        utils.find('div[onClick]').at(1).simulate('click');
        const sortMock = wrapper.find('ResultUtils').first().prop('changeSorting');
         expect(sortMock.mock.calls.length).toBe(2);
        expect(sortMock.mock.calls[0][0]).toBe('release_date');
        expect(sortMock.mock.calls[1][0]).toBe('vote_average');
    });
});

describe('mapStateToProps function for ResultUtils', () => {
    it('returns an object with props mapped from state values', () => {
        const fakeState = {
            movies: {
                results: [1,2,3,4,9],
                sort: 'xxxyyy'
            },
            heading: {
                search: 'may of some year',
                by: 'date',
                selected: mockMovies[0]
            }
        };
        expect(mapStateToProps(fakeState)).toEqual({
            numFound: 5,
            currentSort: 'xxxyyy',
            search: 'may of some year',
            by: 'date',
            selected: mockMovies[0]
        });
    });
});

describe('mapDispatchToProps function for ResultUtils', () => {
    beforeAll(()=>{
        global.fetch = jest.fn();
        global.fetch.mockImplementation(fetchMock);
    });
    it('creates changeSorting function', () => {
        expect(mapDispatchToProps()).toEqual({
            changeSorting: expect.any(Function)
        });
    });
    it('changeSorting function dispatches SORT_MOVIES action with given criterion and result of API call', async () => {
        const fakeDispatch = jest.fn();
        const props = mapDispatchToProps(fakeDispatch);
        await props.changeSorting('crit', 'search', 'by');
        expect(fakeDispatch).toHaveBeenCalledTimes(1);
        expect(fakeDispatch.mock.calls[0][0]).toEqual(sort_movies('crit', `${swaggerBase}/movies?search=search&searchBy=by&sortBy=crit&sortOrder=desc`));
    })
});
