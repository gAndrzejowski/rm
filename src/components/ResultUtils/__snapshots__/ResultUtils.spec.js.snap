// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultUtil component renders correctly with movie not selected 1`] = `
<div
  className={undefined}
>
  <p>
    10 movies found
  </p>
  <div
    className={undefined}
  >
    <div>
      Sort By
    </div>
    <div
      className={undefined}
      onClick={[Function]}
      role="button"
    >
      release date
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      rating
    </div>
  </div>
</div>
`;

exports[`ResultUtil component renders correctly with movie selected 1`] = `
<div
  className={undefined}
>
  <p>
    Films by Crime genre
  </p>
  <div
    className={undefined}
  >
    <div>
      Sort By
    </div>
    <div
      className={undefined}
      onClick={[Function]}
      role="button"
    >
      release date
    </div>
    <div
      className=""
      onClick={[Function]}
      role="button"
    >
      rating
    </div>
  </div>
</div>
`;
